Adding loopTool to analysis tools (#1098)
watertap-org/watertap: docs/technical_reference/tools/loopTool.rst

Loop Tool
====================

.. index::
   pair: watertap.tools.analysis_tools.loop_tool;loop_tool

.. currentmodule:: watertap.tools.analysis_tools.loop_tool

> The loopTool is a wrapper for the parameter sweep (PS) tool set, and is designed to simplify setting up parametric sweeps, enabling sweeping over discreet design choices, and providing a structured data management.  

Reviewer: ```suggestion
The loopTool is a wrapper for the parameter sweep (PS) tool set, and is designed to simplify setting up parametric sweeps, enabling sweeping over discrete design choices, and providing structured data management.  
```

Setting up loopTool
-----------------------------
The loopTool setup involves creating the following:
1)	A flowsheet with build, initialize (optimal), and optimize function
2)	A .yaml file with simulation configuration 
3)	A .py file that connects looptool to the flowsheet, directories for where to save data, and yaml file, as well as other solve or PS options 

Setting up a flowsheet for use with loopTool
----------------------------------------------------
The loopTool requires similar functional setup as PS tool kit. Here we will setup RO_with_energy_recovery.py exmaple flowsheet for use with PS and loopTool.The RO_with_energy_recovery has an option 
that pass in erd_type allowing user to select type of ERD device (either a pressure_exchanger, pump_as_turbine, or None). The user likely would want to run parameter sweep across these options, and we wil setup the loopTool to do so. 

> *The user will need to setup the following three function, description and example are shown below:*

Reviewer: ```suggestion
*The user will need to setup the following three functions (description and example are shown below):*
```

> 1. The *build_function* that builds the flowsheet (ro_build)- This function should build the flowsheet, and accept any kwargs for its configuration. Here we pass explicitly erd_type, but also include *kwargs* incase we want to pass in another options in the future. When loopTool runs, it will pass selectd *kwargs* into this function before running the sweep. 

Reviewer: ```suggestion
1. The *build_function* that builds the flowsheet (ro_build)- This function should build the flowsheet, and accept any kwargs for its configuration. Here we pass explicitly erd_type, but also include *kwargs* in case we want to pass in other options in the future. When loopTool runs, it will pass selected *kwargs* into this function before running the sweep. 
```

> 2. The *initialization_function* (ro_init)- this function will initialize the model. This is in general optional, as initialization can be done in build function as well, but it is useful for PS tool, as it can use it to reinitialize the model in case a solve fails and try again. 

Reviewer: ```suggestion
2. The *initialization_function* (ro_init)- this function will initialize the model. This is optional, as initialization can be done in build function as well, but it is useful for the Parameter Sweep tool, as it can use it to reinitialize the model in case a solve fails and try again. 
```

>    Additionally, the PS tool (and loopTool) will update the parameters on model tree we are sweep across before calling initialize as well as before calling the solve function. It is critical that in initialize function, we fix or unfix any variable we are sweeping over to their required values for initialize routine unless, that routine can leverage changes in sweeped parameters. 

Reviewer: ```suggestion
   Additionally, the Parameter Sweep tool (and loopTool) will update the parameters on the model tree that is swept across before calling initialize as well as before calling the solve function. It is critical that in the initialize function, we fix or unfix any variables we are sweeping over to their required values for the initialization routine, unless that routine can leverage changes in sweep parameters. 
```

>    In our planned sweep, we will be sweeping across flow_mass_phase_comp of NaCl, and RO water recovery. For initialization, these two variables need to be unfixed, so we unfix them at start of ro_init function, as they will be fixed before by PS tool.

Reviewer: ```suggestion
   In our planned sweep, we will be sweeping across flow_mass_phase_comp of NaCl, and RO water recovery. For initialization. These two variables need to be unfixed, so we unfix them at the start of the ro_init function, as they will have been fixed previously by the Parameter Sweep tool.
```

         |  |-outputs 
         |  |-solve_successful 
         |  |-sweep_params 
         |
         |-factor_membrane_replacement # contains standard PS h5 output
         |  |-outputs 
         |  |-solve_successful 
         |  |-sweep_params 
         |
         |-map_sweep # contains standard PS h5 output
            |-outputs 
            |-solve_successful 
            |-sweep_params 

We can readely access the data for processing by useing h5py. 
To visually explore the h5 file, use HDFviewer ( https://www.hdfgroup.org/downloads/hdfview/ )
All the parameters will use thier parm names or object keys as reference (e.g. if you set up yaml file to sweep over 'RO_recovery' for which param is 'fs.ro.ro_recovery', the file will store data for RO_recovery under key 'fs.ro.ro_recovery'.

.. code-block::

   import h5py
    h5file = h5py.File(
        "ro_with_erd_analysisType_ro_erd_type_analysis.h5", "r"
    )

    # get data for cost from  pressure_exchanger erd device, not the dir path strucure
    data = h5file[
        "ro_erd_type_analysis/erd_type/pressure_exchanger/membrane_cost/outputs/fs.costing.LCOW/value"
    ][()] # wil create a list

**Backup management**

> The loopTool includes a naïve data management schema to prevent overwriting existing files and minimize simulations runs through creation of backups from existing files. When the loopTool starts it will check if a file with same name already exists, if it does it will rename that file to include a data and time (file_name+_M_D-H_M-S.bak).  

Reviewer: ```suggestion
The loopTool includes a naïve data management schema to prevent overwriting existing files and minimize simulation runs through creation of backups from existing files. When the loopTool starts it will check if a file with same name already exists, if it does it will rename that file to include a data and time (file_name+_M_D-H_M-S.bak).  
```
